File: python/51-100/95. Unique Binary Search Trees II.py
# Definition for a binary tree node.

class Solution(object):
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        def helper(s, e):
            res = []
            if s > e:
                res.append(None)
                return res
            if s == e:
                res.append(TreeNode(s))
                return res
            for i in range(s, e + 1):
                left = helper(s, i - 1)
                right = helper(i + 1, e)
                for leftnode in left:
                    for rightnode in right:
                        root = TreeNode(i)
                        root.left = leftnode
                        root.right = rightnode
                        res.append(root)
            return res
        return helper(1, n)

File: cpp/51-100/95. Unique Binary Search Trees II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	vector<TreeNode*> generateTrees(int n) {
	    vector<TreeNode *> res;
	    if(n == 0)
	    return res;
		return generator(1, n);
	}
	vector<TreeNode *> generator(int s, int e) {
		vector<TreeNode *> res;
		if (s > e) {
		    res.push_back(NULL);
			return res;
		}
		for (int i = s; i <= e; i++) {
			vector<TreeNode *> left_nodes = generator(s, i - 1);
			vector<TreeNode *> right_nodes = generator(i + 1, e);
			for(auto left: left_nodes)
				for(auto right: right_nodes)
				{
					TreeNode *root = new TreeNode(i);
					root->left = left;
					root->right = right;
					res.push_back(root);
				}
		}
		return res;
	}
};